File: tree/others/min_max_val_from _BST.cpp
#include<iostream>
using namespace std;

struct BSTNode
{
    int data;
    BSTNode* left;
    BSTNode* right;
};

BSTNode* root;

BSTNode* GetNewNode(int data)
{   
    BSTNode *temp = new BSTNode;  
    temp->data = data;
    temp->left = NULL;
    temp->right = NULL;
    return temp;
};


BSTNode* insert(BSTNode* root, int data)
{
    if (root==NULL)
    {
        root = GetNewNode(data);
        return root;
    }

    if (data<=root->data) 
    {
        root->left = insert(root->left, data);
    }
    else
    {
        root->right = insert(root->right, data);
    }

    return root;
}

int min_from_BST(BSTNode* root)
{
    if (root==NULL)
    {
        return 0;
    }
    
    if (root->left==NULL)
    {
        return root->data;
    }

    min_from_BST(root->left);
    
    
}

int max_from_BST(BSTNode* root)
{
    if (root==NULL)
    {
        return 0;
    }
    
    if (root->right==NULL)
    {
        return root->data;
    }

    max_from_BST(root->right);
}

int main()
{
    root=NULL;
    cout<<"Implement Binary Tree"<<endl;
    int n;
    cout<<"Enter n value to be enter in tree ";
    cin>>n;
    for (int i = 0; i < n; i++)
    {
        int data;
        cout<<"Enter data in new node ";
        cin>>data;
        root = insert(root, data);
        cout<<endl;
        cout<<"root data "<<root->data<<endl;
        cout<<endl;
    }

    int min_val = min_from_BST(root);
    int max_val = max_from_BST(root);
    cout<<"min value "<<min_val<<endl;
    cout<<"max value "<<max_val;
    return 0;
}

File: tree/others/min_max_val_from _BST.py
class Node:
    def __init__(self, val):
        self.left = None
        self.right = None
        self.key = val

def insert(root, val):
    if root == None:
        root = Node(val)
        return root

    if val < root.key:
        root.left = insert(root.left, val)
    else:
        root.right = insert(root.right, val)
    return root

def min_from_BST(root):
    if root is None:
        return 0
    if root.left is None:
        return root.key 
    a = min_from_BST(root.left)

    return a

# def min_from_BST(root):
#     if root is None:
#         return float('inf')
#     res = root.key
#     lres = min_from_BST(root.left)
#     rres = min_from_BST(root.right)
#     if lres < res:
#         res = lres
#     if rres < res:
#         res = rres
        
#     return res

def max_from_BST(root):
    if root==None:
        return 0
    if root.right == None:
        return root.key
    a=max_from_BST(root.right)
    return a

if __name__=="__main__":
    root = None

    print("Enter n value in Tree")
    n = int(input())
    for i in range(n):
        print("Enter value ")
        val = int(input())
        root = insert(root, val)

    min_val = min_from_BST(root)
    max_val = max_from_BST(root)

    print("min val from BST", min_val)
    print("max val from BST", max_val)